src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +417,247 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+            array_view->buffer_views[1].data.as_int32 + offset_plus_length;
+        src_view.size_bytes = sizeof(int32_t);
+
+        dst_view.data.data = &(array_view->buffer_views[2].size_bytes);
+        dst_view.size_bytes = sizeof(int32_t);
+
+        // Note that this strategy (copying four bytes into an int64_t)
+        // only works on little-endian (should be checked at a higher level).
+        array_view->buffer_views[2].size_bytes = 0;
+        NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaBufferCopyAsync(
+            device, src_view, ArrowDeviceCpu(), dst_view, hstream));
+      }
+
+      break;
+
+    case NANOARROW_TYPE_LARGE_STRING:
+    case NANOARROW_TYPE_LARGE_BINARY:
+      if (array_view->buffer_views[2].size_bytes == -1) {
+        src_view.data.as_int64 =
+            array_view->buffer_views[1].data.as_int64 + offset_plus_length;
+        src_view.size_bytes = sizeof(int64_t);
+
+        dst_view.data.data = &(array_view->buffer_views[2].size_bytes);
+        dst_view.size_bytes = sizeof(int64_t);
+
+        array_view->buffer_views[2].size_bytes = 0;
+        NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaBufferCopyAsync(
+            device, src_view, ArrowDeviceCpu(), dst_view, hstream));
+      }
+      break;
+    default:
+      break;
+  }
+
+  // Recurse for children
+  for (int64_t i = 0; i < array_view->n_children; i++) {
+    NANOARROW_RETURN_NOT_OK(
+        ArrowDeviceCudaResolveBufferSizesAsync(device, array_view->children[i], hstream));
+  }
+
+  // ...and for dictionary
+  if (array_view->dictionary != NULL) {
+    NANOARROW_RETURN_NOT_OK(
+        ArrowDeviceCudaResolveBufferSizesAsync(device, array_view->dictionary, hstream));
+  }
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaArrayViewCopyBuffers(
+    struct ArrowDevice* device_src, struct ArrowArrayView* src,
+    struct ArrowDevice* device_dst, struct ArrowArray* dst, CUstream hstream,
+    int64_t* unknown_buffer_size_count) {
+  // Currently no attempt to minimize the amount of memory copied (i.e.,
+  // by applying offset + length and copying potentially fewer bytes)
+  dst->length = src->length;
+  dst->offset = src->offset;
+  dst->null_count = src->null_count;
+
+  for (int i = 0; i < NANOARROW_MAX_FIXED_BUFFERS; i++) {
+    if (src->layout.buffer_type[i] == NANOARROW_BUFFER_TYPE_NONE) {
+      break;
+    }
+
+    struct ArrowBufferView src_view = src->buffer_views[i];
+    int64_t buffer_size = src_view.size_bytes;
+
+    // No need to initialize a buffer of zero bytes
+    if (buffer_size == 0) {
+      continue;
+    }
+
+    // buffer_size will be -1 if it is unknown (i.e., string/binary
+    // data buffer when copying from CUDA -> CPU)
+    if (buffer_size < 0) {
+      (*unknown_buffer_size_count)++;
+      continue;
+    }
+
+    struct ArrowBuffer* buffer_dst = ArrowArrayBuffer(dst, i);
+
+    // Don't initialize a buffer that has already been initialized
+    if (buffer_dst->size_bytes > 0) {
+      continue;
+    }
```

Comment:
This ensures a buffer (that was already copied) doesn't get copied a second time when calling `ArrowDeviceCudaArrayViewCopyBuffers` a second time?